test/unit/queue/consumers/ManagedConsumer-test.js

```diff
@@ -0,0 +1,199 @@
+
+const EventEmitter = require('events')
+
+const waitForEvent = require('~/test/util/waitForEvent')
+
+const TEST_AMQ_URL = 'amq-url'
+const TEST_QUEUE_URL = 'test-queue'
+
+// simple mock queue consumer
+class MockQueueConsumer extends EventEmitter {
+  start () {
+    return Promise.resolve()
+  }
+
+  stop () {
+    return Promise.resolve()
+  }
+}
+
+test.beforeEach('setup mock channel and connections', (t) => {
+  const sandbox = sinon.sandbox.create()
+
+  const simpleConsumer = new MockQueueConsumer()
+
+  const ManagedConsumer = proxyquire('~/queue/ManagedConsumer', {
+    './util/createConsumer': () => simpleConsumer
+  })
+
+  t.context = { sandbox, simpleConsumer, ManagedConsumer }
+})
+
+test.afterEach('clean up', (t) => {
+  const { sandbox } = t.context
+  sandbox.restore()
+})
+
+// tests
+test('should be able to create a consumer with "start"', async (t) => {
+  const { ManagedConsumer } = t.context
+
+  const managedConsumer = new ManagedConsumer({
+    amqUrl: TEST_AMQ_URL,
+    logger: console,
+    restart: true,
+    consumerOptions: {
+      queueName: TEST_QUEUE_URL
+    },
+    onMessage: Promise.resolve
+  })
+
+  await managedConsumer.start()
+
+  t.truthy(managedConsumer.getConsumer())
+})
+
+test('should not return a consumer if not started', async (t) => {
+  const { ManagedConsumer } = t.context
+
+  const managedConsumer = new ManagedConsumer({
+    amqUrl: TEST_AMQ_URL,
+    logger: console,
+    restart: true,
+    consumerOptions: {
+      queueName: TEST_QUEUE_URL
+    },
+    onMessage: Promise.resolve
+  })
+
+  await managedConsumer.start()
+
+  t.truthy(managedConsumer.getConsumer())
+})
+
+test('should recreate a connection for a new consumer after ' +
+'a consumer errors out', async (t) => {
+  const { ManagedConsumer, simpleConsumer } = t.context
+
+  const managedConsumer = new ManagedConsumer({
+    amqUrl: TEST_AMQ_URL,
+    logger: console,
+    restart: true,
+    reconnectTimeout: 1,
+    consumerOptions: {
+      queueName: TEST_QUEUE_URL
+    },
+    onMessage: Promise.resolve
+  })
+
+  await managedConsumer.start()
+
+  const restartPromise = Promise.all([
+    waitForEvent(managedConsumer, 'consumer-stopped'),
+    waitForEvent(managedConsumer, 'consumer-restarted')
+  ])
+
+  // simulate failure
+  simpleConsumer.emit('error')
+
+  await restartPromise
+  t.pass()
+})
+
+test('should not recreate a connection for a new consumer after ' +
+'a consumer errors out (connection lost)', async (t) => {
+  const { ManagedConsumer, simpleConsumer } = t.context
+
+  const managedConsumer = new ManagedConsumer({
+    amqUrl: TEST_QUEUE_URL,
+    logger: console,
+    restart: false,
+    reconnectTimeout: 1,
+    consumerOptions: {
+      queueName: TEST_QUEUE_URL
+    },
+    onMessage: Promise.resolve
+  })
+
+  await managedConsumer.start()
+
+  const restartPromise = Promise.resolve()
+    .then(() => waitForEvent(managedConsumer, 'consumer-restarted'))
+    .timeout(500)
```

Comment:
Should be able to use sinon.useFakeTimers for these tests. Will save us a lot of time in the long run if we avoid waiting for these timeouts.